File: py/portfolio.py
from money import Money
from functools import reduce
import operator

class Portfolio:
    def __init__(self):
        self.moneys = []

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                convertedMoney = bank.convert(m, currency)
                total += convertedMoney.amount
            except Exception as ke:
                failures.append(ke)

        if not failures:
            return Money(total, currency)

        failureMessage = ",".join(f.args[0] for f in failures)
        raise Exception('Missing exchange rate(s):[' + failureMessage + ']')

File: js/portfolio.js
const Money = require('./money')

class Portfolio{
    constructor(){
        this.moneys = [];
    }

    add(...moneys){
        this.moneys = this.moneys.concat(moneys)
    }

    convert(money, currency){
        let exchageRates = new Map();
        exchageRates.set('EUR->USD', 1.2);
        exchageRates.set('USD->KRW', 1100);

        if (money.currency === currency) {
            return money.amount
        }
        let key = money.currency + '->' + currency;
        let rate = exchageRates.get(key)
        if (rate === undefined){
            return undefined
        }
        return money.amount * rate;
    }

    evaluate(bank, currency){
        let failures = [];
        let total = this.moneys.reduce( (sum, money) => {
            try{
                let convertedMoney = bank.convert(money, currency)
                return sum + convertedMoney.amount
            } catch (error) {
                failures.push(error.message);
                return sum;
            }
        }, 0)
        if (!failures.length) {
            return new Money(total, currency)
        }
        throw new Error("Missing exchange rate(s):[" + failures.join() + "]");
    }
}

module.exports = Portfolio;